File: Tests/PUPSISPortalTests/ScheduleParserTests.swift
import XCTest
@testable import PUPSISPortal

/// Cases are the real schedule lines scraped from the live SIS page.
final class ScheduleParserTests: XCTestCase {
    private func parse(_ line: String) -> [ClassSession] {
        ScheduleParser.parse([
            "subjectCode": "TEST 001",
            "description": "Test Subject",
            "faculty": "SANTOS, JUAN",
            "scheduleLine": line,
        ])
    }

    func testTwoDaysPairWithTheirOwnTimes() {
        let sessions = parse("1N - BSCS 1-1N - T/F 02:00PM-04:00PM/01:30PM-04:30PM")
        XCTAssertEqual(sessions.count, 2)
        XCTAssertEqual(sessions[0].day, .tuesday)
        XCTAssertEqual(sessions[0].start, 14 * 60)
        XCTAssertEqual(sessions[0].end, 16 * 60)
        XCTAssertEqual(sessions[1].day, .friday)
        XCTAssertEqual(sessions[1].start, 13 * 60 + 30)
        XCTAssertEqual(sessions[1].end, 16 * 60 + 30)
    }

    func testSameDayTwiceBecomesTwoBlocks() {
        let sessions = parse("1N - BSCS 1-1N - SUN/SUN 08:00AM-12:00PM/01:00PM-06:00PM")
        XCTAssertEqual(sessions.count, 2)
        XCTAssertTrue(sessions.allSatisfy { $0.day == .sunday })
        XCTAssertEqual(sessions[0].start, 8 * 60)
        XCTAssertEqual(sessions[0].end, 12 * 60)
        XCTAssertEqual(sessions[1].start, 13 * 60)
        XCTAssertEqual(sessions[1].end, 18 * 60)
    }

    func testSingleDaySingleRange() {
        let sessions = parse("1N - BSCS 1-1N - S 07:30AM-10:30AM")
        XCTAssertEqual(sessions.count, 1)
        XCTAssertEqual(sessions[0].day, .saturday)
        XCTAssertEqual(sessions[0].start, 7 * 60 + 30)
        XCTAssertEqual(sessions[0].end, 10 * 60 + 30)
    }

    /// `TH` and `SUN` must win over the single-letter `T`/`S` codes.
    func testMultiLetterDayCodesArentSwallowed() {
        XCTAssertEqual(parse("1N - X - TH 09:00AM-11:00AM").first?.day, .thursday)
        XCTAssertEqual(parse("1N - X - SUN 09:00AM-11:00AM").first?.day, .sunday)
    }

    func testNoonAndMidnightBoundaries() {
        let noon = parse("1N - X - M 12:00PM-01:00PM")
        XCTAssertEqual(noon.first?.start, 12 * 60)
        XCTAssertEqual(noon.first?.end, 13 * 60)

        let earlyMorning = parse("1N - X - M 12:30AM-01:30AM")
        XCTAssertEqual(earlyMorning.first?.start, 30)
        XCTAssertEqual(earlyMorning.first?.end, 90)
    }

    func testUnparseableRowsAreSkippedNotCrashed() {
        XCTAssertTrue(parse("").isEmpty)
        XCTAssertTrue(parse("1N - BSCS 1-1N - TBA").isEmpty)
        XCTAssertTrue(parse("no times here at all").isEmpty)
    }
}

/// `Calendar` counts weekdays from Sunday, this enum counts from Monday.
/// Getting the shift wrong puts the now-line in the wrong column.
final class WeekdayTests: XCTestCase {
    private func weekday(_ iso: String) throws -> Weekday {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = try XCTUnwrap(TimeZone(identifier: "Asia/Manila"))

        var formatter = DateComponents()
        let parts = iso.split(separator: "-").compactMap { Int($0) }
        (formatter.year, formatter.month, formatter.day) = (parts[0], parts[1], parts[2])
        formatter.hour = 12

        return Weekday.on(try XCTUnwrap(calendar.date(from: formatter)), calendar: calendar)
    }

    func testEveryDayOfAKnownWeekMaps() throws {
        // 2026-08-03 is a Monday.
        XCTAssertEqual(try weekday("2026-08-03"), .monday)
        XCTAssertEqual(try weekday("2026-08-04"), .tuesday)
        XCTAssertEqual(try weekday("2026-08-05"), .wednesday)
        XCTAssertEqual(try weekday("2026-08-06"), .thursday)
        XCTAssertEqual(try weekday("2026-08-07"), .friday)
        XCTAssertEqual(try weekday("2026-08-08"), .saturday)
        XCTAssertEqual(try weekday("2026-08-09"), .sunday)
    }

    /// `Calendar.firstWeekday` is locale-dependent — in a US locale the week
    /// starts Sunday, which would slide the whole grid by a day.
    func testWeekStartIsMondayRegardlessOfLocale() throws {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = try XCTUnwrap(TimeZone(identifier: "Asia/Manila"))
        calendar.firstWeekday = 1 // Sunday, as en_US would have it

        let monday = try date("2026-08-03", calendar)

        for iso in ["2026-08-03", "2026-08-05", "2026-08-09"] {
            let start = Weekday.weekStart(containing: try date(iso, calendar), calendar: calendar)
            XCTAssertEqual(start, monday, "week containing \(iso)")
        }
    }

    /// Sunday belongs to the week that started six days earlier, not the one
    /// about to begin.
    func testSundayClosesTheWeekRatherThanOpeningTheNext() throws {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = try XCTUnwrap(TimeZone(identifier: "Asia/Manila"))

        let sunday = try date("2026-08-09", calendar)
        let start = Weekday.weekStart(containing: sunday, calendar: calendar)

        XCTAssertEqual(start, try date("2026-08-03", calendar))
        XCTAssertEqual(Weekday.sunday.date(inWeekStarting: start, calendar: calendar), sunday)
    }

    func testEachWeekdayLandsOnItsOwnDate() throws {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = try XCTUnwrap(TimeZone(identifier: "Asia/Manila"))
        let start = try date("2026-08-03", calendar)

        let dates = Weekday.allCases.map { $0.date(inWeekStarting: start, calendar: calendar) }

        XCTAssertEqual(dates.count, Set(dates).count)
        XCTAssertEqual(dates.first, start)
        XCTAssertEqual(dates.last, try date("2026-08-09", calendar))
    }

    private func date(_ iso: String, _ calendar: Calendar) throws -> Date {
        let parts = iso.split(separator: "-").compactMap { Int($0) }
        var components = DateComponents()
        (components.year, components.month, components.day) = (parts[0], parts[1], parts[2])
        return try XCTUnwrap(calendar.date(from: components))
    }
}

/// `ClassSession.nextMeetingDate` — drives the class-note Add-date menu's
/// "Next class" default.
final class ClassSessionNextMeetingTests: XCTestCase {
    private var calendar: Calendar = {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "Asia/Manila")!
        return c
    }()

    private func date(_ iso: String) throws -> Date {
        let parts = iso.split(separator: "-").compactMap { Int($0) }
        var components = DateComponents()
        (components.year, components.month, components.day) = (parts[0], parts[1], parts[2])
        return try XCTUnwrap(calendar.date(from: components))
    }

    private func session(_ subject: String, _ day: Weekday) -> ClassSession {
        ClassSession(subjectCode: subject, description: "", faculty: "", day: day, start: 480, end: 600)
    }

    /// Subject meets Sunday; asking from Saturday (the 15th) should land on
    /// Sunday the 16th, not today.
    func testWrapsForwardToTheNextScheduledWeekday() throws {
        let sessions = [session("COMP1", .sunday)]
        let saturday = try date("2026-08-15")

        let next = ClassSession.nextMeetingDate(for: "COMP1", in: sessions, from: saturday, calendar: calendar)

        XCTAssertEqual(next, try date("2026-08-16"))
    }

    /// Asking on the meeting day itself returns that same day, not a week out.
    func testReturnsTodayWhenTodayIsAMeetingDay() throws {
        let sessions = [session("COMP1", .sunday)]
        let sunday = try date("2026-08-16")

        let next = ClassSession.nextMeetingDate(for: "COMP1", in: sessions, from: sunday, calendar: calendar)

        XCTAssertEqual(next, sunday)
    }

    func testNilForASubjectNotOnTheSchedule() throws {
        let sessions = [session("COMP1", .sunday)]

        let next = ClassSession.nextMeetingDate(for: "MATH1", in: sessions, from: try date("2026-08-15"), calendar: calendar)

        XCTAssertNil(next)
    }
}

File: windows/PUPSISPortal.Core.Tests/ScheduleParserTests.cs
using PUPSISPortal.Core;

namespace PUPSISPortal.Core.Tests;

/// <summary>
/// Cases are the real schedule lines scraped from the live SIS page.
/// </summary>
public class ScheduleParserTests
{
    private static List<ClassSession> Parse(string line)
    {
        return ScheduleParser.Parse(new Dictionary<string, string>
        {
            { "subjectCode", "TEST 001" },
            { "description", "Test Subject" },
            { "faculty", "SANTOS, JUAN" },
            { "scheduleLine", line },
        });
    }

    [Fact]
    public void TwoDaysPairWithTheirOwnTimes()
    {
        var sessions = Parse("1N - BSCS 1-1N - T/F 02:00PM-04:00PM/01:30PM-04:30PM");
        Assert.Equal(2, sessions.Count);
        Assert.Equal(Weekday.Tuesday, sessions[0].Day);
        Assert.Equal(14 * 60, sessions[0].Start);
        Assert.Equal(16 * 60, sessions[0].End);
        Assert.Equal(Weekday.Friday, sessions[1].Day);
        Assert.Equal(13 * 60 + 30, sessions[1].Start);
        Assert.Equal(16 * 60 + 30, sessions[1].End);
    }

    [Fact]
    public void SameDayTwiceBecomesTwoBlocks()
    {
        var sessions = Parse("1N - BSCS 1-1N - SUN/SUN 08:00AM-12:00PM/01:00PM-06:00PM");
        Assert.Equal(2, sessions.Count);
        Assert.True(sessions.All(s => s.Day == Weekday.Sunday));
        Assert.Equal(8 * 60, sessions[0].Start);
        Assert.Equal(12 * 60, sessions[0].End);
        Assert.Equal(13 * 60, sessions[1].Start);
        Assert.Equal(18 * 60, sessions[1].End);
    }

    [Fact]
    public void SingleDaySingleRange()
    {
        var sessions = Parse("1N - BSCS 1-1N - S 07:30AM-10:30AM");
        Assert.Single(sessions);
        Assert.Equal(Weekday.Saturday, sessions[0].Day);
        Assert.Equal(7 * 60 + 30, sessions[0].Start);
        Assert.Equal(10 * 60 + 30, sessions[0].End);
    }

    /// <summary>
    /// `TH` and `SUN` must win over the single-letter `T`/`S` codes.
    /// </summary>
    [Fact]
    public void MultiLetterDayCodesArentSwallowed()
    {
        var th = Parse("1N - X - TH 09:00AM-11:00AM");
        Assert.NotEmpty(th);
        Assert.Equal(Weekday.Thursday, th.First().Day);

        var sun = Parse("1N - X - SUN 09:00AM-11:00AM");
        Assert.NotEmpty(sun);
        Assert.Equal(Weekday.Sunday, sun.First().Day);
    }

    [Fact]
    public void NoonAndMidnightBoundaries()
    {
        var noon = Parse("1N - X - M 12:00PM-01:00PM");
        Assert.NotEmpty(noon);
        Assert.Equal(12 * 60, noon.First().Start);
        Assert.Equal(13 * 60, noon.First().End);

        var earlyMorning = Parse("1N - X - M 12:30AM-01:30AM");
        Assert.NotEmpty(earlyMorning);
        Assert.Equal(30, earlyMorning.First().Start);
        Assert.Equal(90, earlyMorning.First().End);
    }

    [Fact]
    public void UnparseableRowsAreSkippedNotCrashed()
    {
        Assert.Empty(Parse(""));
        Assert.Empty(Parse("1N - BSCS 1-1N - TBA"));
        Assert.Empty(Parse("no times here at all"));
    }
}

/// <summary>
/// DayOfWeek numbers weekdays from Sunday = 0, this enum counts from Monday = 1.
/// Getting the shift wrong puts the now-line in the wrong column.
/// </summary>
public class WeekdayTests
{
    private static Weekday GetWeekday(int year, int month, int day)
    {
        var date = new DateTime(year, month, day, 12, 0, 0);
        return WeekdayExtensions.On(date);
    }

    [Fact]
    public void EveryDayOfAKnownWeekMaps()
    {
        // 2026-08-03 is a Monday.
        Assert.Equal(Weekday.Monday, GetWeekday(2026, 8, 3));
        Assert.Equal(Weekday.Tuesday, GetWeekday(2026, 8, 4));
        Assert.Equal(Weekday.Wednesday, GetWeekday(2026, 8, 5));
        Assert.Equal(Weekday.Thursday, GetWeekday(2026, 8, 6));
        Assert.Equal(Weekday.Friday, GetWeekday(2026, 8, 7));
        Assert.Equal(Weekday.Saturday, GetWeekday(2026, 8, 8));
        Assert.Equal(Weekday.Sunday, GetWeekday(2026, 8, 9));
    }

    /// <summary>
    /// The week should always start on Monday regardless of the system's locale.
    /// </summary>
    [Fact]
    public void WeekStartIsMondayRegardlessOfLocale()
    {
        var monday = new DateTime(2026, 8, 3); // Monday
        var wednesday = new DateTime(2026, 8, 5);
        var sunday = new DateTime(2026, 8, 9);

        var startFromMonday = WeekdayExtensions.WeekStart(monday);
        var startFromWednesday = WeekdayExtensions.WeekStart(wednesday);
        var startFromSunday = WeekdayExtensions.WeekStart(sunday);

        Assert.Equal(monday.Date, startFromMonday);
        Assert.Equal(monday.Date, startFromWednesday);
        Assert.Equal(monday.Date, startFromSunday);
    }

    /// <summary>
    /// Sunday belongs to the week that started six days earlier, not the one
    /// about to begin.
    /// </summary>
    [Fact]
    public void SundayClosesTheWeekRatherThanOpeningTheNext()
    {
        var sunday = new DateTime(2026, 8, 9);
        var monday = new DateTime(2026, 8, 3);

        var start = WeekdayExtensions.WeekStart(sunday);
        Assert.Equal(monday.Date, start);

        var sundayInWeek = Weekday.Sunday.DateInWeekStarting(start);
        Assert.Equal(sunday.Date, sundayInWeek.Date);
    }

    [Fact]
    public void EachWeekdayLandsOnItsOwnDate()
    {
        var start = new DateTime(2026, 8, 3); // Monday
        var dates = new List<DateTime>();
        foreach (Weekday day in Enum.GetValues(typeof(Weekday)))
        {
            dates.Add(day.DateInWeekStarting(start));
        }

        // All dates should be unique
        Assert.Equal(dates.Count, new HashSet<DateTime>(dates).Count);

        // First should be Monday (2026-08-03)
        Assert.Equal(new DateTime(2026, 8, 3), dates.First());

        // Last should be Sunday (2026-08-09)
        Assert.Equal(new DateTime(2026, 8, 9), dates.Last());
    }
}

/// <summary>
/// ClassSession.NextMeetingDate — drives the class-note Add-date menu's
/// "Next class" default.
/// </summary>
public class ClassSessionNextMeetingTests
{
    private static DateTime DateOnly(int year, int month, int day) => new DateTime(year, month, day);

    private static ClassSession Session(string subject, Weekday day) =>
        new ClassSession
        {
            SubjectCode = subject,
            Description = "",
            Faculty = "",
            Day = day,
            Start = 480,
            End = 600
        };

    /// <summary>
    /// Subject meets Sunday; asking from Saturday (the 15th) should land on
    /// Sunday the 16th, not today.
    /// </summary>
    [Fact]
    public void WrapsForwardToTheNextScheduledWeekday()
    {
        var sessions = new List<ClassSession> { Session("COMP1", Weekday.Sunday) };
        var saturday = DateOnly(2026, 8, 15);

        var next = WeekdayExtensions.NextMeetingDate("COMP1", sessions, saturday);

        Assert.Equal(DateOnly(2026, 8, 16), next);
    }

    /// <summary>
    /// Asking on the meeting day itself returns that same day, not a week out.
    /// </summary>
    [Fact]
    public void ReturnsTodayWhenTodayIsAMeetingDay()
    {
        var sessions = new List<ClassSession> { Session("COMP1", Weekday.Sunday) };
        var sunday = DateOnly(2026, 8, 16);

        var next = WeekdayExtensions.NextMeetingDate("COMP1", sessions, sunday);

        Assert.Equal(sunday, next);
    }

    [Fact]
    public void NilForASubjectNotOnTheSchedule()
    {
        var sessions = new List<ClassSession> { Session("COMP1", Weekday.Sunday) };

        var next = WeekdayExtensions.NextMeetingDate("MATH1", sessions, DateOnly(2026, 8, 15));

        Assert.Null(next);
    }
}
